docs: Add CI/CD example for building Argo Workflows using Argo Workflows - Fixes #8591

wesleyscholl opened the pull request:
<!-- markdownlint-disable MD041 -- this is rendered within existing HTML, so allow starting without an H1 -->
## A detailed example for implementing CI/CD pipelines to build Argo Workflows using Argo Workflows
<!--
### When you open your PR
- Once builds are green, mark your PR "Ready for review".
When changes are requested, please address them and then dismiss the review to get it reviewed again.
-->


Fixes #8591 

### Motivation

This issue was more than 2 years old and I wanted to contribute to the project. I work with Argo Workflows everyday and I'm CAPA certified (Certified Argo Project Associate). To prepare for my next DevOps role, I wanted to learn more about CI/CD pipelines and lifecycle.

### Slide Deck from the Community Meeting Demo
<img width="350" alt="Screenshot 2024-12-20 at 5 06 21 PM" src="https://github.com/user-attachments/assets/a27bd83d-8a29-49b1-8401-124876d29f5b" />


[Argo Workflows CI_CD Demo.pdf](https://github.com/user-attachments/files/18215821/Argo.Workflows.CI_CD.Demo.pdf)



### Modifications

`WorkflowTemplate` example includes:

#### CI
| Description                                           | Resource                | Status |
|----------------------------------------------|-------------------------|--------|
| Argo Workflows API endpoint, submits `WorkflowTemplate`   | `WorkflowEventBinding`  | :ballot_box_with_check:     |
| Build the CLI                               | `WorkflowTemplate`      | :ballot_box_with_check:      |
| Build images using Buildkit - Push images to Docker Hub | `WorkflowTemplate` | :ballot_box_with_check:     |
| Run unit tests and collect test report      | `WorkflowTemplate`      | :ballot_box_with_check:     |
| Run coverage and collect report             | `WorkflowTemplate`      | :ballot_box_with_check:     |
| Deploy to a cluster                         | `WorkflowTemplate`      | :ballot_box_with_check:     |
| Run basic E2E tests and collect report      | `WorkflowTemplate`      | :ballot_box_with_check:     |

#### Pipeline 
|  Description                                     | Resource                | Status |
|----------------------------------------------|-------------------------|--------|
| Combine CI and CD jobs with an approval step | `WorkflowTemplate`      | :ballot_box_with_check:     |

#### CD
| Description                                            | Resource                | Status |
|----------------------------------------------|-------------------------|--------|
| Tag and push tag                            | `WorkflowTemplate`      | :ballot_box_with_check:     |
| Update deployment manifests using `kustomize edit set image` | `WorkflowTemplate` | :ballot_box_with_check:     |
| Commit deployment manifests                 | `WorkflowTemplate`      | :ballot_box_with_check:     |
| Start Argo CD sync step                     | `WorkflowTemplate`      | :ballot_box_with_check:     |


#### Documentation & Configuration
|  Description                                     | Resource                | Status |
|----------------------------------------------|-------------------------|--------|
| CI, CD, & Pipeline Documentation   |  Comments    | :ballot_box_with_check:     |
| Service Account for `WorkflowEventBinding`   |  `ServiceAccount`   | :ballot_box_with_check:     |
| `$ARGO_TOKEN` - Bearer Authorization for `WorkflowEventBinding` | `Secret` |   :ballot_box_with_check:     |
| Role to submit workflow templates for `WorkflowEventBinding` | `Role` |   :ballot_box_with_check:     |
|`RoleBinding` the `Role` to the `ServiceAccount` for `WorkflowEventBinding`|`RoleBinding`| :ballot_box_with_check: |
| Docker configuration - Hub credentials | `Secret` |   :ballot_box_with_check:     |
| GitHub PAT - GitHub credentials | `Secret` |   :ballot_box_with_check:     |
| ArgoCD server and credentials | `Secret` |   :ballot_box_with_check:     |
| ArgoCD network policy  | `NetworkPolicy` |   :ballot_box_with_check:     |

### Verification

#### Thoroughly tested the workflow:
- [x] E2E Workflow
<img width="1908" alt="Screenshot 2024-12-17 at 4 33 40 PM" src="https://github.com/user-attachments/assets/4da22afa-d58e-4d28-b27b-4aa87c383668" />

- [x] Workflow steps

- `clone-repo` 
<img width="749" alt="Screenshot 2024-11-07 at 9 29 08 AM" src="https://github.com/user-attachments/assets/6444e576-16bb-4a4c-93f4-e55efa29e466">

- `build-cli`
<img width="750" alt="Screenshot 2024-11-07 at 9 50 37 AM" src="https://github.com/user-attachments/assets/a6c46889-528c-404b-9f8e-5e2fd37abd9c">


- `create-exec-image`
<img width="1097" alt="Screenshot 2024-11-07 at 10 13 17 AM" src="https://github.com/user-attachments/assets/b40e2688-e1b5-42a8-b0dc-1c6c5dbecb11">


- `create-cli-image`
<img width="1099" alt="Screenshot 2024-11-07 at 10 55 04 AM" src="https://github.com/user-attachments/assets/a46cbc1b-8fc4-4a39-9e4f-927791d25720">


- `run-tests`
<img width="958" alt="Screenshot 2024-11-07 at 12 02 53 PM" src="https://github.com/user-attachments/assets/1582a054-4e96-40c6-b637-06377e934167">


- `run-coverage`
<img width="1392" alt="Screenshot 2024-11-07 at 12 09 01 PM" src="https://github.com/user-attachments/assets/89a41ccc-6f4c-46b1-9139-86ea9788e4d9">


- `prepare-deploy-to-cluster-e2e-test`
<img width="853" alt="Screenshot 2024-11-12 at 11 35 58 AM" src="https://github.com/user-attachments/assets/646a4b38-e8a5-4522-8aac-f0a5dd0b0144">


- `approval`
<img width="1009" alt="Screenshot 2024-11-08 at 1 00 57 PM" src="https://github.com/user-attachments/assets/ae528e8f-ac88-44c5-b4eb-2c4bdfef47bc">


- `docker-tag-push`
<img width="787" alt="Screenshot 2024-11-08 at 1 04 11 PM" src="https://github.com/user-attachments/assets/923212e1-2180-403a-a4c0-751f077a97b8">


- `update-manifests`
<img width="749" alt="Screenshot 2024-11-08 at 1 04 46 PM" src="https://github.com/user-attachments/assets/be539063-5a13-481c-95e4-1e3144d86f20">


- `commit-manifests`
<img width="758" alt="Screenshot 2024-11-08 at 1 05 08 PM" src="https://github.com/user-attachments/assets/d8078884-2734-4508-966d-e727c8e078b4">


- `start-argocd-sync`
<img width="1448" alt="Screenshot 2024-11-08 at 1 07 01 PM" src="https://github.com/user-attachments/assets/2df1af1d-4311-4fda-85af-d06d4b0dab90">

wesleyscholl committed:
docs: Add CI/CD example for building Argo Workflows using Argo Workflows - Fixes #8591

Signed-off-by: Wesley Scholl <128409641+wesleyscholl@users.noreply.github.com>

wesleyscholl committed:
Merge branch 'main' into ci-cd-example-workflow

wesleyscholl committed:
docs: Add apiVersion and WorkflowTemplate kind to CI/CD example - Fixes #8591

Signed-off-by: Wesley Scholl <128409641+wesleyscholl@users.noreply.github.com>

wesleyscholl committed:
docs: Add apiVersion and WorkflowTemplate kind to CI/CD example workflow - Fixes #8591

Signed-off-by: Wesley Scholl <128409641+wesleyscholl@users.noreply.github.com>

wesleyscholl committed:
Merge branch 'main' into ci-cd-example-workflow

wesleyscholl commented:
For review: @alexec @terrytangyuan @sarabala1979 @caelan-io 

Thanks

terrytangyuan reviewed (commented):
This is great but my main concern is how difficult it is to maintain this in this repo

mubarak-j commented:
I'm not sure how this adds any complexity or burden when there are more than 50 examples in this repo. The lack of documentation/example for CI/CD use case, makes this a much-needed addition, as described in the linked issue.

wesleyscholl commented:
@terrytangyuan 

The example is a bit long, but it includes all the requirements from #8591. Would be it easier to maintain if it was split into separate CI and CD `WorkflowTemplates`? Then link them in the docs.

@Joibel 

Thanks for reviewing, this is the first CI/CD pipeline I've ever built. I want to modify this example to make it fully functional. For instance, adding this line to build the controller:

```shell
make controller kit STATIC_FILES=false
```
The workflow can implement a DAG to run all of the e2e tests in parallel and use the `argo-cli` build in the e2e tests.

As for the scope and caveats, should more comments be added to the example to explain each step in the workflow?

wesleyscholl committed:
Merge branch 'main' into ci-cd-example-workflow

wesleyscholl committed:
docs(examples): Add comprehensive test matrix and parallel execution support - Fixes #8591

- Added test matrix to cover all E2E test cases.
- Introduced a DAG for parallel test execution to optimize runtime.
- Updated workflow steps for improved parallelism.
- Refactored code to use generic templates, reducing duplication and overall code length.
- Enhanced code readability by adding detailed comments.
- Added build step for the controller.

Signed-off-by: Wesley Scholl <128409641+wesleyscholl@users.noreply.github.com>

wesleyscholl committed:
docs(examples): Add CI/CD example for building Argo Workflows - Fixes #8591

Signed-off-by: Wesley Scholl <128409641+wesleyscholl@users.noreply.github.com>

wesleyscholl committed:
Merge branch 'main' into ci-cd-example-workflow

wesleyscholl commented:
@terrytangyuan @Joibel @mubarak-j @alexec 

This PR has been updated:

- [x] Added test matrix to cover all E2E test cases.
- [x] Introduced a DAG for parallel test execution to optimize runtime.
- [x] Updated workflow steps for improved parallelism.
- [x] Refactored code to use generic templates, reducing duplication and overall code length.
- [x] Enhanced code readability by adding detailed comments.
- [x] Added build step for the controller.

sarabala1979 commented:
I am kind of thinking to create argoproj-lab repo to keep e2e CI/CD configuration for end user like product which implemented on top argo workflow.
WDYT?
@terrytangyuan

sarabala1979 commented:
@wesleyscholl Thanks for contributing this usecase. But just examble is not enough.  Community is looking full fetch e2e product. You demo this in upcoming community meeting.   Do you like to contribute?

terrytangyuan commented:
>I am kind of thinking to create argoproj-lab repo to keep e2e CI/CD configuration for end user like product which implemented on top argo workflow.

Sure as long as there's a way to maintain and test it continuously going forward.

wesleyscholl commented:
@sarabala1979 @terrytangyuan

I can demo for the upcoming community meeting (12/18) and I'm interested in contributing to a full-fledged e2e CI/CD product. A CI/CD extension/plugin that integrates with Argo Workflows sounds very interesting. I can also help with the maintenance and tests.

wesleyscholl committed:
Merge branch 'main' into ci-cd-example-workflow

caelan-io commented:
Hi @wesleyscholl - thanks a lot for developing this example! I agree with @mubarak-j that it's definitely needed. Looking forward to reviewing it further myself. 

Happy to have you join us at our last community meeting of the year next Weds, Dec 18 @ 10a PST. 🙌 I have added you to the [agenda](https://docs.google.com/document/d/1DKR4VQ-04DJfHOh6S6-cbQw6MiWppC0k7afORiiy6XQ/edit?pli=1&tab=t.0#heading=h.3ayc915uy5ye) and we will record the meeting and publish to YouTube afterwards. Please feel free to message me in CNCF Slack if you have any questions ahead of the meeting.

caelan-io commented:
An idea that came to mind while seeing @wesleyscholl 's presentation today at the Community Meeting: We could create a new docs page for "CI/CD use cases" and add links to various examples for users to check out and try on their own. 

👍 if you'd like to see this page created and we can suggest some content to add to it from community meetings, examples like this, blog posts, and the Pipekit library of free resources. 

I agree that we probably want to improve upon this example before it's blessed as the canonical CI/CD example in the docs, but I think it can serve as a good example for others to take inspiration from for their own CI/CD pipelines using Argo Workflows at their companies, hobby projects, etc. Maybe, like @sarabala1979 recommended, it's hosted in argoproj-labs and referenced in the docs as an unofficial example?

terrytangyuan commented:
There's an existing page for CI/CD here: https://argo-workflows.readthedocs.io/en/latest/use-cases/ci-cd/

wesleyscholl commented:
@terrytangyuan @caelan-io 

Should I add a link to the CI/CD use cases docs page?

terrytangyuan commented:
>Should I add a link to the CI/CD use cases docs page?

Yes that would be great

wesleyscholl committed:
docs(workflow): Update test matrix software versions and improve step configuration

- Updated test matrix to reflect latest software versions for compatibility testing.
- Added detailed comments to improve workflow maintainability.
- Set workingDir for all steps to ensure consistent file paths.
- Adjusted resource requests and limits (CPU/memory) to align with step requirements and runtime optimization.
- Refactored step/task commands to conform to WorkflowTemplate format for modularity.
- Combined unit tests and test coverage into a single step for efficiency.
- Improved readability and reduced redundancy by centralizing template logic.
- Added link in ci cd use cases docs to example workflow.

Signed-off-by: Wesley Scholl <128409641+wesleyscholl@users.noreply.github.com>

wesleyscholl commented:
@terrytangyuan @caelan-io

I added a link to the CI/CD use cases docs page. Let me know if you need anything else, thanks.

wesleyscholl committed:
Merge branch 'main' into ci-cd-example-workflow
